opentracing-shim/src/opentracingshim/__init__.py

```diff
@@ -0,0 +1,241 @@
+                self._tracer, otel_span.get_context(), otel_span
+            )
+            yield ScopeWrapper(self, wrapped_span)
+
+    @property
+    def active(self):
+        span = self._tracer.get_current_span()
+        wrapped_span = SpanWrapper(self._tracer, span.get_context(), span)
+        return ScopeWrapper(self, wrapped_span)
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object (and the same for ScopeWrapper?).
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class TracerWrapper(opentracing.Tracer):
+    def __init__(self, tracer, scope_manager=None):
+        self._otel_tracer = tracer
+
+        # If a scope manager isn't provided by the user, create a
+        # `ScopeManagerWrapper` instance and use it to initialize the
+        # `TracerWrapper`.
+        if scope_manager is None:
+            scope_manager = ScopeManagerWrapper(self)
+
+        super().__init__(scope_manager=scope_manager)
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `Tracer` object."""
+
+        return self._otel_tracer
+
+    @property
+    def active_span(self):
+        span = self._otel_tracer.get_current_span()
+        if span is None:
+            return None
+        return SpanWrapper(self, span.get_context(), span)
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+    @contextmanager
+    def start_active_span(
+        self,
+        operation_name,
+        child_of=None,
+        references=None,
+        tags=None,
+        start_time=None,
+        ignore_active_span=False,
+        finish_on_close=True,
+    ):
+        span = self.start_span(
+            operation_name=operation_name,
+            child_of=child_of,
+            references=references,
+            tags=tags,
+            start_time=start_time,
+            ignore_active_span=ignore_active_span,
+        )
+        with self._scope_manager.activate(span, finish_on_close) as scope:
+            yield scope
+
+    def start_span(
+        self,
+        operation_name=None,
+        child_of=None,
+        references=None,
+        tags=None,
+        start_time=None,
+        ignore_active_span=False,
+    ):
+        parent = child_of
+        if parent is not None:
+            if isinstance(parent, SpanWrapper):
+                parent = child_of.unwrap()
+            elif isinstance(parent, SpanContextWrapper):
+                parent = child_of.unwrap()
+            else:
+                raise RuntimeError(
+                    "Invalid parent type when calling start_span()."
+                )
+            # TODO: Should we use the OpenTracing base classes for the type
+            # check?
+
+        # Use active span as parent when no explicit parent is specified.
+        if (
```

Comment:
Instead of doing this here, just put `if child_of is None and not ignore_active_span: child_of = self.active_span` at the beginning of this function, then you can reuse the unwrap/check/etc logic.

Reply:
@Oberon00 I've tried refactoring according to your suggestion but I don't end up with simpler code. Could you describe the full logic (pseudocode is fine)? Maybe I am missing your point.

Please keep in mind we need to address the following cases as far as I can tell:

- Valid parent specified.
- Invalid parent specified, we can use the active span as parent.
- Invalid parent specified, we can't use the active span as parent (because there is no active span or because `ignore_active_span` is `True`).
- No parent specified, we can use the active span as parent.
- No parent specified, we can't use the active span as parent.

Reply:
This is what I mean:
```python

def start_span(
        # ...
    ):
        # Use active span as parent when no explicit parent is specified.
        if (
            not ignore_active_span
            and not child_of
        ):
            child_of = self.active_span

        parent = child_of
        if parent is not None:
            # ... as before ... but without the old `if`.

Reply:
@Oberon00 the problem is that I wanted to use the active span as parent also when `child_of` contains a value of an invalid type, and log a warning about the invalid parent type in this case. The suggestion above doesn't handle this case and I can't seem to find an elegant way to incorporate the type check + log action into the above logic.

Maybe we should just give up this type check, then we can greatly simplify:

```python
def start_span(
    #...
):
   # Use active span as parent when no explicit parent is specified.
   if not ignore_active_span and not child_of:
       child_of = self.active_span

   parent = child_of.unwrap() if child_of else None
   span = self._otel_tracer.create_span(operation_name, parent)
```

I'm starting to think that doing the type check would be taking too much responsibility in the shim. The user who calls the shim's OT API should pass the right types... WDYT?

Reply:
>In that case you currently throw an exception, right?

No, at the moment I log a warning and fall back to using the current span as parent, or using `None` as a parent, which causes the created span to begin a new trace.

>I have no strong opinion on whether we should handle that. But we should probably not throw an exception explicitly (failing with an AttributError when an unexpected type that is outside the API spec is passed is fine with me though).

Yes, this is what will happen if I disable the check and an invalid type is passed in `child_of`.

So, I'm opting for disabling the check. This is also consistent with https://github.com/open-telemetry/opentelemetry-python/pull/211#discussion_r334904577. We can always revisit this at a later stage.